lib/service_check.pm

```diff
@@ -116,14 +123,31 @@ our $default_services = {
     },
 };
 
+#
+# use subroutine to call the function with parameters
+#
+sub call_func {
+    # assuming that $f is a function reference
+    my $f = shift;
+    $f->(@_);
+}
+
 sub install_services {
-    my $service = shift;
+    my ($service) = @_;
+    my $service_func;
     $hdd_base_version = get_var('HDDVERSION');
     foreach my $s (keys %$service) {
         my $srv_pkg_name  = $service->{$s}->{srv_pkg_name};
         my $srv_proc_name = $service->{$s}->{srv_proc_name};
         my $support_ver   = $service->{$s}->{support_ver};
         if (grep { $_ eq $hdd_base_version } split(',', $support_ver)) {
+            # service check before migration, if we've set up service check
+            # function, we don't need following actions to check the service.
+            if (exists $service->{$s}->{service_check_func}) {
+                $service_func = $service->{$s}->{service_check_func};
+                call_func(\&$service_func, 'before');
```

Comment:
This flow looks dangerous to me, as if key is somehow removed, we won't test anything and show green bubble in openQA. I would be more strict about the check.

Also, general side note. With such design, this method will become hell in case we will get even 20 services. I'm not even talking about the performance, that we will have to go through the list of all services to match the name, even though we always know which service exactly we need.

Reply:
OK, for (1) we can add softfail for the missed service functional check.
       for (2) I don't get your point, can you gave me some pseudo code for explanation.
Thanks!

Reply:
So, for second, we have loop which iterates over services list with different conditions:
```
if (exists $service->{$s}->{service_check_func}) {
if ($srv_pkg_name ne 'apparmor') {
```
It already mixes the purpose as function name states that it installs service,  but we already run some checks. `install_services` take a list of services as an argument, and in case we call it for e.g. apache, we will still compare it if it's not apparrmor. However, when we need this check, we know exactly that we need only apache. I suspect that when we will have more service checks, this loop will have many special cases which will be irrelevant most of the time.

In your verification run I don't see http_srv module being scheduled at all, so in case apache configuration fails, it will be hard to debug, because they are executed all at once in the single loop. On top, if something fails for one service, we won't even try to test remaining ones (that's the way openQA works).
So at some point of time, it will be worth splitting it and solving this by schedule, as we know what we need when.
For example, there will be module `apache_fresh_installation` and `apache_post_migration` which will simply reuse methods from the library. Then you don't need to check which service is that as you know what you need for it, instead of applying sophisticated logic to detect if we have not migrated service yet or we already did with 'before' and 'after' magic strings.
Hope it's a bit more clear now.

Reply:
@rwx788 
We have to make specific checks for apparmor since this is workround for a bug, if new bug happen in future maybe we still need more specific check. 
In verification run indeed you can not see http_srv module since we just call the function within http_srv, this will lead difficult for debug indeed.
For solving this by schedule, my understanding is you need create apache_lib.pm in /lib and create apache_fresh_installation.pm and  apache_post_migration.pm in /tests, then we schedule in main.pm, am i correct?
Then the logic(which service support on which version etc, currently in service_check.pm) should move to main.pm , then this will make complex on schedule part(main.pm). Correct me if i am wrong.